### ISSUE TITLE
Complains about dict indexing after assert(thing, dict).

### ISSUE THREAD

**Issue by gvanrossum** [astral-sh/ty#3476]
### Summary

Playground link: https://play.ty.dev/e14be9dc-fb5a-46d3-93d4-32cf9b780d3a

Code:
```py
def func(rows: object) -> object:
    assert isinstance(rows, dict)
    return rows['min_score']
```
Diagnostic:
```
warning[invalid-argument-type]: Method `__getitem__` of type `bound method Top[dict[Unknown, Unknown]].__getitem__(key: Never, /) -> object` cannot be called with key of type `Literal["min_score"]` on object of type `Top[dict[Unknown, Unknown]]`
 --> tt.py:3:12
  |
3 |     return rows['min_score']
  |            ^^^^
  |
```

**Comment by carljm** [comment]
Thanks for the report!

This is a known choice we currently make -- since `isinstance(x, dict)` doesn't tell us anything about the key or value type, we have to choose to either be safe or forgiving about that lack of knowledge. Choosing to use `dict[Any, Any]` or equivalent here (as other type checkers do) means we'd allow `rows[whatever].anything_at_all()` (and many other variants, including mutating the dict by adding entries of any type) without complaint. Choosing to be safe means that this case (narrowing from `object`) may require a `cast` or a `type: ignore` if you want to do much of anything with the dict other than iterate it. (There's no problem with our current behavior in the more common cases, such as narrowing from a type like `dict[str, int] | list[int]` with `isinstance(x, dict)` -- just when narrowing from `object` where we have no prior information at all.)

If we switched to using `dict[Any, Any]` as the narrowing type for `isinstance(x, dict)`, then (given a correct implementation of intersection types) narrowing from `dict[str, int] | list[int]` would not simplify to `dict[str, int]`, it would end up with the type `dict[str, int] & dict[Any, Any]`, which does not simplify, and is probably too forgiving.

I do think we need better options here; #3375 is currently our best idea for how to provide them.

This issue is very similar to #2374, but that one has the added wrinkle that using `.get()` on an unknown dictionary is safer than indexing it directly. I think it makes sense to leave both issues open.

**Comment by carljm** [comment]
> dict[str, int] & dict[Any, Any]

I think this intersection actually could correctly simplify to `dict[str, int]`, given invariance of `dict`, and _if_ we assume that intersections with gradual types don't materialize to `Never` (which was a topic we discussed recently at the typing summit, cc @JelleZijlstra ). I don't think this fully solves this question (since it can also apply to covariant types), but it would make an option to use gradual types for isinstance narrowing more feasible.

**Comment by gvanrossum** [comment]
I looked deeper into where this came from in our code -- it's a rarely-run benchmark where the type is listed as `object`, because the author didn't want to write a complex union type: it's got at least six types, one of which is the `dict[str, float]` that is assumed here -- so I should have described `func` as returning `float`, not `object`, though this doesn't matter here.

I think it's just up to me to fix this, even though other type checkers (at least pyright) allowed it.

**Comment by sharkdp** [comment]
Starting with ty 0.0.69, the original example here doesn't error anymore, since we've switched to a more gradual `isinstance` narrowing behavior, which follows more what other type checkers do. If someone wants to go back to the previous, more strict / more correct behavior, you can enable the [`analysis.strict-generic-narrowing`](https://docs.astral.sh/ty/reference/configuration/#strict-generic-narrowing) setting.
